overlord/snapshotstate/backend/backend.go

```diff
@@ -323,3 +327,150 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 
 	return nil
 }
+
+type exportMetadata struct {
+	Format int       `json:"format"`
+	Date   time.Time `json:"date"`
+	Files  []string  `json:"files"`
+}
+
+// Export allows exporting a given snapshot set
+func Export(ctx context.Context, setID uint64) (*SnapshotExport, error) {
+	return NewSnapshotExport(ctx, setID)
+}
+
+type SnapshotExport struct {
+	// open snapshot files
+	snapshotFiles []*os.File
+
+	// size
+	size int64
+}
+
+// NewSnapshotExport will return a SnapshotExport structure. It must be
+// Close()ed after use to avoid leaking file descriptors.
+func NewSnapshotExport(ctx context.Context, setID uint64) (se *SnapshotExport, err error) {
+	var snapshotFiles []*os.File
+
+	defer func() {
+		// cleanup any open FDs if anything goes wrong
+		if err != nil {
+			for _, f := range snapshotFiles {
+				f.Close()
+			}
+		}
+	}()
+
+	// Open all files first and keep the file descriptors
+	// open. The caller should have locked the state so that no
+	// delete/change snapshot operations can happen while the
+	// files are getting opened.
+	err = Iter(ctx, func(reader *Reader) error {
+		if reader.SetID == setID {
+			// dup() the file descriptor
+			fd, err := syscall.Dup(int(reader.Fd()))
+			if err != nil {
+				return fmt.Errorf("cannot dup %v", reader.Name())
+			}
+			f := os.NewFile(uintptr(fd), reader.Name())
+			if f == nil {
+				return fmt.Errorf("invalid fd %v for %v", fd, reader.Name())
+			}
+			snapshotFiles = append(snapshotFiles, f)
+		}
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("cannot export snapshot %v: %v", setID, err)
+	}
+	if len(snapshotFiles) == 0 {
+		return nil, fmt.Errorf("no snapshot data found for %v", setID)
+	}
+
+	// Export once into a dummy writer so that we can set the size
+	// of the export. This is then used to set the Content-Length
+	// in the response correctly.
+	//
+	// Note that the size of the generated tar could change if the
+	// time switches between this export and the export we stream
+	// to the client to a time after the year 2242. This is unlikely
+	// but a known issue with this approach here.
+	var sz osutil.Sizer
+	se = &SnapshotExport{snapshotFiles: snapshotFiles}
```

Comment:
isn't this kind of expensive? it seems to merit at least a TODO to try to improve on this? looking again maybe is not even too hard, just needs an extra method on Sizer to add to the size without going through an actual Write.

Reply:
I updated the code now so that this happens outside of the state lock. It is expensive and maybe it can be done via a specialized sizer. Not entirely trivial though because we need the final size of the container that is exported (the tar file) not just the sum of the files we find.